engine/classes/Elgg/Ajax/Service.php

```diff
@@ -0,0 +1,200 @@
+<?php
+namespace Elgg\Ajax;
+
+use Elgg\Http\Input;
+use Elgg\PluginHooksService;
+use Elgg\SystemMessagesService;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Elgg\Services\AjaxResponse;
+
+/**
+ * Models the Ajax API service
+ *
+ * @since 1.12.0
+ * @access private
+ * @internal
+ */
+class Service {
+
+	/**
+	 * @var PluginHooksService
+	 */
+	private $hooks;
+
+	/**
+	 * @var SystemMessagesService
+	 */
+	private $msgs;
+
+	/**
+	 * @var Input
+	 */
+	private $input;
+
+	/**
+	 * @var bool
+	 */
+	private $response_sent = false;
+
+	/**
+	 * Constructor
+	 *
+	 * @param PluginHooksService    $hooks Hooks service
+	 * @param SystemMessagesService $msgs  System messages service
+	 * @param Input                 $input Input service
+	 */
+	public function __construct(PluginHooksService $hooks, SystemMessagesService $msgs, Input $input) {
+		$this->hooks = $hooks;
+		$this->msgs = $msgs;
+		$this->input = $input;
+	}
+
+	/**
+	 * Did the request come from the elgg/ajax module?
+	 *
+	 * @return bool
+	 */
+	public function isAjax2Request() {
+		$version = _elgg_services()->request->headers->get('X-Elgg');
```

Comment:
We might want to check the HMAC headers used in web services to avoid possible conflicts.